plugins/vpp/abfplugin/descriptor/abf.go

```diff
@@ -0,0 +1,281 @@
+		Retrieve:             d.Retrieve,
+		IsRetriableFailure:   d.IsRetriableFailure,
+		DerivedValues:        d.DerivedValues,
+		Dependencies:         d.Dependencies,
+		RetrieveDependencies: []string{ifdescriptor.InterfaceDescriptorName},
+	}
+}
+
+// EquivalentABFs compares related ACL name, list of attached interfaces and forwarding paths to
+// specify ABS equality.
+func (d *ABFDescriptor) EquivalentABFs(key string, oldABF, newABF *abf.ABF) bool {
+	// check index and associated ACL
+	if oldABF.AclName != newABF.AclName {
+		return false
+	}
+
+	// compare attached interfaces
+	if len(oldABF.AttachedInterfaces) != len(newABF.AttachedInterfaces) {
+		return false
+	}
+	if !equivalentABFAttachedInterfaces(oldABF.AttachedInterfaces, newABF.AttachedInterfaces) {
+		return false
+	}
+
+	// compare forwarding paths
+	if len(oldABF.ForwardingPaths) != len(newABF.ForwardingPaths) {
+		return false
+	}
+	return equivalentABFForwardingPaths(oldABF.ForwardingPaths, newABF.ForwardingPaths)
+}
+
+// Validate validates VPP ABF configuration.
+func (d *ABFDescriptor) Validate(key string, abfData *abf.ABF) error {
+	if abfData.Index == "" {
+		return api.NewInvalidValueError(ErrABFInvalidIndex, "index")
+	}
+	_, err := strconv.Atoi(abfData.Index)
+	if err != nil {
+		return api.NewInvalidValueError(ErrABFInvalidIndex, "index")
+	}
+	if abfData.AclName == "" {
+		return api.NewInvalidValueError(ErrABFWithoutACL, "acl_name")
+	}
+	return nil
+}
+
+// Create validates ABF (mainly index), verifies ACL existence and configures ABF policy. Attached interfaces
+// are put to metadata together with the ABF index to make it available for other ABF descriptors.
+func (d *ABFDescriptor) Create(key string, abfData *abf.ABF) (*abfidx.ABFMetadata, error) {
+	abfIdx, err := getABFIndex(abfData)
+	if err != nil {
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// get ACL index
+	aclData, exists := d.aclIndex.LookupByName(abfData.AclName)
+	if !exists {
+		err := errors.Errorf("failed to obtain metadata for ACL %s", abfData.AclName)
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// add new ABF policy
+	if err := d.abfHandler.AddAbfPolicy(uint32(abfIdx), aclData.Index, abfData.ForwardingPaths); err != nil {
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// fill the metadata
+	metadata := &abfidx.ABFMetadata{
+		Index:    uint32(abfIdx),
+		Attached: abfData.AttachedInterfaces,
+	}
+
+	return metadata, nil
+}
+
+// Delete removes ABF policy
+func (d *ABFDescriptor) Delete(key string, abfData *abf.ABF, metadata *abfidx.ABFMetadata) error {
+	// ACL ID is not required
+	return d.abfHandler.DeleteAbfPolicy(metadata.Index, abfData.ForwardingPaths)
+}
+
+// UpdateWithRecreate is always set to true since there is no binary API to specialy handle a part
+// of the config - the whole ABF policy needs to be removed and created.
+func (d *ABFDescriptor) UpdateWithRecreate(key string, oldAbfData, newAbfData *abf.ABF, oldMetadata *abfidx.ABFMetadata) bool {
+	// always recreate
```

Comment:
UpdateWithRecreate can be omitted - without Update method, the re-creation is automatic (but keep it if it is for the comment).